### PR TITLE
[WIP] amp-fx-collection fade-in

### PR DESCRIPTION
**Pull request by nainar**
This is a WIP PR. It adds the following functionality:
* adds the ability to accept `amp-fx="fade-in"` which triggers a timed fade in animation 
* customizing the fade in animation by specifying:
  * duration of animation
  * easing curve of the animation
  * margin at which the animation should be triggered

This PR needs a UX review as of yet. 
Sending out for technical review to expedite.

### CODE REVIEW

**Comment by nainar** [comment]
Adding @aghassemi to cc as well. And @alanorozco .

File: examples/amp-fx-collection.html [hunk 1]
```diff
@@ -0,0 +1,319 @@
+<!--
+
+ _______   ______      .__   __.   ______   .___________.        _______. __    __  .______   .___  ___.  __  .___________.
```
Comment (alanorozco) [inline examples/amp-fx-collection.html]: :)

File: extensions/amp-fx-collection/0.1/providers/fade-in.js [hunk 2]
```diff
@@ -0,0 +1,205 @@
+/**
```
Comment (alanorozco) [inline extensions/amp-fx-collection/0.1/providers/fade-in.js]: Haven't given a detailed view at this, but it looks to me like most of this can be abstracted out? Is there a reason why we can't describe the animations themselves in a more declarative way?

File: examples/amp-fx-collection.html [hunk 3]
```diff
@@ -0,0 +1,319 @@
+<!--
```
Comment (nainar) [inline examples/amp-fx-collection.html]: Might submit this as an example actually. Without the large DO NOT SUBMIT ASCII art that is

File: extensions/amp-fx-collection/0.1/providers/fade-in.js (hunk 2 above)
Comment (nainar) [reply extensions/amp-fx-collection/0.1/providers/fade-in.js]: I'm sorry I don't understand the question here, could you rephrase it?

Comment (alanorozco) [reply extensions/amp-fx-collection/0.1/providers/fade-in.js]: Sure.

It seems to me like the provider does the following:

1. Gathers configuration parameters and sets default values.
2. Calculates relative position for triggering, etc.
3. Applies interpolation (`update_`)

These semantics will apply to most, if not all, fx presets.

If we're writing a provider for each of them, we will have some duplicated code and it will probably be a little tough to maintain them.

My question really is whether something is blocking you from writing a generic `FxProvider` that simply takes in an interpolation function and/or a default installer. Similar to the API in `animation.js`.

File: examples/amp-fx-collection.html (hunk 3 above)
Comment (alanorozco) [reply examples/amp-fx-collection.html]: sg!

File: extensions/amp-fx-collection/0.1/providers/fade-in.js (hunk 2 above)
Comment (nainar) [reply extensions/amp-fx-collection/0.1/providers/fade-in.js]: Ahhh gotcha. Can definitely do that! Thanks for the explanation.

**Comment by nainar** [comment]
@alanorozco - I have refactored as the code as you suggested. PTAL? Thanks!

File: extensions/amp-fx-collection/0.1/providers/fade-in.js [hunk 4]
```diff
@@ -14,43 +14,16 @@
...
-
-    /** @private @const {!../../../../src/service/resources-impl.Resources} */
-    this.resources_ = Services.resourcesForDoc(ampdoc);
-
-    installPositionObserverServiceForDoc(ampdoc);
-
-    /** @private @const {!../../../../src/service/position-observer/position-observer-impl.PositionObserver} */
-    this.positionObserver_ = getServiceForDoc(ampdoc, 'position-observer');
-  }
+export class FadeInProvider extends FxProvider {
```
Comment (alanorozco) [inline extensions/amp-fx-collection/0.1/providers/fade-in.js]: Looking better!

One nit, though. This is a classic example of composition over inheritance. If we keep from allocating new classes for each fx, then we're making a big leap in performance and maintanability.

How about passing the interpolation function to the FxProvider constructor? That way you also only have to write tests for `FxProvider` but not for subclasses (since they wouldn't exist).

If you need more granular control over the provider then consider a mixin approach, where each Fx would be a root class composed into `FxProvider`.

Comment (nainar) [reply extensions/amp-fx-collection/0.1/providers/fade-in.js]: I don't understand how passing the function into a constructor would happen in this case. The interpolation function (`update_()` in this case) needs access to variables stored on the classes themselves.

Comment (alanorozco) [reply extensions/amp-fx-collection/0.1/providers/fade-in.js]: 1. update() can take in the animation configuration params
2. Instead of interpolation being a function, it could be a class.
